### DISCUSSION TITLE
Param: xyz can only be a request body, using Body() when using discriminated unions

### DISCUSSION THREAD

**Discussion by DurandA** [fastapi/fastapi#9761] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Body
from typing import Annotated, Union, Literal

from pydantic import Field, BaseModel

class ItemBase(BaseModel):
    item_type: str
    item_value: list | int | str

class ScaleItem(ItemBase):
    item_type: Literal['scale']
    item_value: int

class FreeFormItem(ItemBase):
    item_type: Literal['free_form']
    item_value: str

class GenericItem(BaseModel):
    __root__: Union[ScaleItem, FreeFormItem] = Field(discriminator="item_type")

GenericItemWithoutRoot = Annotated[Union[ScaleItem, FreeFormItem], Field(discriminator="item_type")]

app = FastAPI()

@app.post("/item")
async def process_item(
    foo: GenericItemWithoutRoot,
):
    return {"Hello": "World"}
```

### Description

I am using `Annotated` with a [discriminated `Union`](https://docs.pydantic.dev/1.10/usage/types/#discriminated-unions-aka-tagged-unions).

When creating a route with the discrimated `Union` (`GenericItemWithoutRoot` in the example) as input and expecting it to be the request body, the following error is thrown:
> AssertionError: Param: foo can only be a request body, using Body()

However, I can neither force `foo` to be a body using `foo: GenericItemWithoutRoot = Body()` (_AssertionError: Cannot specify FastAPI annotations in `Annotated` and default value together for 'foo'_) or using `foo: Annotated[GenericItemWithoutRoot, Body(embed=True)]` (_AssertionError: Cannot specify multiple `Annotated` FastAPI arguments for 'foo'_).

The only workaround I found is to use a model with `__root__` (`GenericItem`) which is [not recommended](https://github.com/tiangolo/fastapi/discussions/9709#discussioncomment-6256511).

### Operating System

Linux

### FastAPI Version

0.97.0

### Python Version

Python 3.11.4

**Comment by odiseo0** [comment]
I had the same problem, my solution was to change `Field` for `Body`

```py
from fastapi import FastAPI, Body
from typing import Annotated, Union, Literal

from pydantic import Field, BaseModel

class ItemBase(BaseModel):
    item_type: str
    item_value: list | int | str

class ScaleItem(ItemBase):
    item_type: Literal['scale']
    item_value: int

class FreeFormItem(ItemBase):
    item_type: Literal['free_form']
    item_value: str

class GenericItem(ItemBase):
    __root__: Union[ScaleItem, FreeFormItem] = Field(discriminator="item_type")

GenericItemWithoutRoot = Annotated[Union[ScaleItem, FreeFormItem], Body()]

app = FastAPI()

@app.post("/item")
async def process_item(
    foo: GenericItemWithoutRoot,
):
    return {"Hello": "World"}
```

**Reply by DurandA** [reply]
Thank you for posting your solution. I am pleasantly surprised that it works without an explicit discriminator. I tried with `return {"Hello": type(foo).__name__}` and it always returned the right type, even if item_value was set to the same type for all classes.

I however see at least one limitation, the resulting OpenAPI definition uses [_anyOf_ insead of _oneOf_](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/).

Btw, there was a mistake in my original sample code:
```patch
- class GenericItem(ItemBase):
+ class GenericItem(BaseModel):
    __root__: Union[ScaleItem, FreeFormItem] = Field(discriminator="item_type")
```

**Comment by DurandA** [comment]
I think the appropriate fix would be to support multiple annotated value, e.g. `GenericItemWithoutRoot = Annotated[Union[ScaleItem, FreeFormItem], Body(), Field(discriminator="item_type")]`. This is however explicitly prevented by FastAPI:
> _AssertionError: Cannot specify multiple `Annotated` FastAPI arguments for 'foo'_

**Reply by haoyun** [reply]
I am facing the same problem and I hope this discussion will draw more attention.

**Reply by odiseo0** [reply]
```py
GenericItemWithoutRoot = Annotated[Union[ScaleItem, FreeFormItem], Body(discriminator="item_type")]
```
Works since the 0.101.0 I believe

**Reply by Rendez** [reply]
This works for me:

```
GenericItemWithoutRoot = Annotated[
    Union[
        ScaleItem,
        FreeFormItem,
    ],
    Body(...),
]
```

**Reply by haoyun** [reply]
The reason that I don't quite like this solution is that it makes my models depend on FastAPI. However, I would like to decouple my models and the fastapi app. Moreover, if the disjoint union comes from a 3rd party package, then I have no way to change from `Field(...)` to `Body(...)`.

**Reply by psychedelicious** [reply]
@DurandA @haoyun This issue is resolved for us by #10773

**Comment by YuriiMotov** [comment, marked as answer]
Seems that it has been fixed and works fine in current version (fastapi==0.115.12, pydantic==2.11.5)
